Shared/ViewModels/DevicesViewModel.swift

```diff
@@ -0,0 +1,241 @@
+//
+// Swiftfin is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2024 Jellyfin & Jellyfin Contributors
+//
+
+import Combine
+import Foundation
+import JellyfinAPI
+import OrderedCollections
+import SwiftUI
+
+final class DevicesViewModel: ViewModel, Stateful {
+
+    // MARK: - Action
+
+    enum Action: Equatable {
+        case getDevices
+        case setCustomName(id: String, newName: String)
+        case deleteDevice(id: String)
+        case deleteAllDevices
+    }
+
+    // MARK: - BackgroundState
+
+    enum BackgroundState: Hashable {
+        case gettingDevices
+        case settingCustomName
+        case deletingDevice
+        case deletingAllDevices
+    }
+
+    // MARK: - State
+
+    enum State: Hashable {
+        case content
+        case error(JellyfinAPIError)
+        case initial
+    }
+
+    @Published
+    final var backgroundStates: OrderedSet<BackgroundState> = []
+    @Published
+    final var devices: OrderedDictionary<String, BindingBox<DeviceInfo?>> = [:]
+    @Published
+    final var state: State = .initial
+
+    private var deviceTask: AnyCancellable?
+
+    // MARK: - Respond to Action
+
+    func respond(to action: Action) -> State {
+        switch action {
+        case .getDevices:
+            deviceTask?.cancel()
+
+            deviceTask = Task { [weak self] in
+                await MainActor.run {
+                    let _ = self?.backgroundStates.append(.gettingDevices)
```

Comment:
Could this be simplified a little by adding it to the states before creating the task?